### ISSUE TITLE
[ENH] Next steps for development on experimental surface module

### ISSUE THREAD

**Issue by ymzayek** [nilearn/nilearn#4027]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe your proposed enhancement in detail.

Following the merge of #3856, here's a non exhaustive list of follow-up tasks for the surface API:

Maskers:
- From (https://github.com/nilearn/nilearn/pull/3856#issuecomment-1723692075):
  - improve a bit the masker for the very frequent use case where we don't want to mask anything by avoiding to store the mask img 
  - Implement a `SurfaceVectorizer` that just checks shapes and concatenates data but doesn't have a mask_img and doesn't mask 
- Implement extraction of signals from ROI and transforming region signals back to surface object

Decoder: 
- Update `_check_embedded_nifti_masker` to work with surface masker

Plotting:
- Adapting surf plotting functions for `SurfaceImage`

Documentation:
- Add information about surface API to user guide
- Add surface terms to glossary

### Benefits to the change

Continue development of new module

### Pseudocode for the new behavior, if applicable

```python
# insert your code below

```

**Comment by ymzayek** [comment]
List of (old) issues and PRs to address or to close once integration is done and stable:
- #2682 
- #2424 
- #3265 
- #3846

**Comment by Remi-Gau** [comment]
Copying some of the notes from our hackathon regarding writing surface data:

Writing surface images

- Separate meshes and textures into different gifti files
- Question about naming: should try and make reasonable defaults for that, yet leave the users the possibility to change names
  - Suggestion to go with with something BIDS like where the hemisphere:
    - https://bids-specification.readthedocs.io/en/latest/derivatives/introduction.html#gifti-surface-data-format 
  - [hemi-{L|R}][_space-<space>][_den-<label>]_suffix.surf.gii
    - Wonder how much of those we can infer without having to keep track of the name of the input file…
    - Suffix is going to be almost impossible to have a sensible default for
    - Hemi specifies the hemisphere of a given file
    - Space for example “fsaverage”
    - Den for density, for example:
      - Den-10k for "10242 vertices per hemisphere (5th order icosahedron)"

**Comment by ymzayek** [comment]
@Remi-Gau we need to note that #4120 needs to be reverted if we release without the surface API.

In addition I'm thinking whether it would be better to develop the integration on a feature branch?

**Comment by Remi-Gau** [comment]
> @Remi-Gau we need to note that #4120 needs to be reverted if we release without the surface API.

Good point

> In addition I'm thinking whether it would be better to develop the integration on a feature branch?

We could try that indeed: to avoid those reversions, it would make sense. 
Just need to make sure that we back merge main into it often to avoid diverging (though I suspect that if we work mostly on experimental stuff then we should be fine)

**Comment by bthirion** [comment]
What's preventing from releasing with the Surface API ? The plotting functions ?

**Comment by jeromedockes** [comment]
> @Remi-Gau we need to note that https://github.com/nilearn/nilearn/pull/4120 needs to be reverted if we release without the surface API.

maybe we could avoid reverting it. to avoid importing the experimental module we can add some attribute for duck typing or even just check the class name instead of the `isinstance` checks in the `check_embedded_nifti_masker` and decoding functions. Once the module is no longer experimental we can change it back

**Comment by Remi-Gau** [comment]
As far as I can tell the SurfaceMasker does not include a `generate_report` method like the other maskers.

**Comment by ymzayek** [comment]
> As far as I can tell the SurfaceMasker does not include a `generate_report` method like the other maskers.

I define it and raise NotImplemetedError in #4126

**Comment by ymzayek** [comment]
Oh  meant that's for GLM but I think you can do the same for the masker for now

**Comment by jeromedockes** [comment]
Yasmin ***@***.***> writes:

>> As far as I can tell the SurfaceMasker does not include a `generate_report` method like the other maskers.
>
> I define it and raise NotImplemetedError in #4126
>

Why is that better than not defining it?

**Comment by ymzayek** [comment]
@jeromedockes to raise a message with the error that it is not yet implemented  and imply that it will be. But feel free to do otherwise

**Comment by Remi-Gau** [comment]
> Why is that better than not defining it?

I would say mostly as signal to users (and future us) that this is coming.

**Comment by Remi-Gau** [comment]
met with @bthirion and @jeromedockes at the open office hours

Decisions:

- need to better develop the experimental aspects of the Surface features before we move it into the stable API
  - integrate plotting first
  - integrate surface GLM 
  - integrate decoding
- so the next release will be a minor release instead of a major one

**Comment by bthirion** [comment]
... and for the time being, we represent meshes through dictionaries.

**Comment by Remi-Gau** [comment]
@bthirion @htwangtw @emdupre @alexisthual @ymzayek 

Opinions needed

@jeromedockes and I chatted a bit more regarding the next steps.

Currently on main, the experimental module is imported in the following "non-experiemntal" places.

```
$ find nilearn -name '*.py' -exec grep -H "nilearn.experimental" {} \;

nilearn/decoding/space_net.py:from nilearn.experimental.surface import SurfaceMasker
nilearn/decoding/decoder.py:from nilearn.experimental.surface import SurfaceMasker
nilearn/decoding/tests/test_decoder.py:from nilearn.experimental.surface import SurfaceMasker
nilearn/_utils/masker_validation.py:from nilearn.experimental.surface import SurfaceMasker
nilearn/_utils/tests/test_masker_validation.py:from nilearn.experimental.surface import SurfaceMasker
```

This leads to the warning from experimental module to be thrown when users use non-experimental part of the codebase.

https://github.com/nilearn/nilearn/blob/cfda9ff95426181940104e5119597b795bce938d/nilearn/experimental/__init__.py#L10

We feel that rather reverting #4120 (see https://github.com/nilearn/nilearn/issues/4027#issuecomment-1888823619) it would be OK to get rid of the experimental warning, assuming that users wanting to use the experimental features would literal have to type "experimental". 

This could be replaced with a additional warning in the doc: https://nilearn.github.io/dev/modules/experimental.html and in the coming release notes.

How do others feel about this?

**Comment by Remi-Gau** [comment]
TODO:

- [ ] modified examples from #4126 should be copied as a new experimental example and the the "old" example should mention the new experimental example

A similar approach (copy to experimental + mention new experimental example in old version) should be be used for other examples where the new surface API can be applied.

- [ ] update top message of this issue with a list of examples to be modified with the new surface API

**Comment by jeromedockes** [comment]
> would literal have to type "experimental".

indeed that is more likely to catch their attention than any warning

**Comment by jeromedockes** [comment]
> ... and for the time being, we represent meshes through dictionaries.

or rather collections of meshes containing one or more hemispheres. each individual mesh is still represented by a `Mesh` instance

**Comment by jeromedockes** [comment]
> I would say mostly as signal to users (and future us) that this is coming.

ok my 2 cents is stating it exists by defining it and then surprise it doesn't when we call it is not super helpful, and not defining it at all is less work but it's not important either way

**Comment by Remi-Gau** [comment]
Adding a section in the top messages to mention issues or PR that will be made obsolete or superseded by the ones created here

**Comment by Remi-Gau** [comment]
> * BIDS compatibility: make it possible for first_level_from_bids to return surface images directly

will update some of data generation to facilitate writing tests for this

**Comment by Remi-Gau** [comment]
update of https://github.com/nilearn/nilearn/issues/4027#issuecomment-1908667900

```
nilearn/reporting/glm_reporter.py:from nilearn.experimental.surface import SurfaceMasker
nilearn/decoding/space_net.py:from nilearn.experimental.surface import SurfaceMasker
nilearn/decoding/decoder.py:from nilearn.experimental.surface import SurfaceMasker
nilearn/glm/first_level/first_level.py:from nilearn.experimental.surface import SurfaceImage, SurfaceMasker
nilearn/_utils/masker_validation.py:from nilearn.experimental.surface import SurfaceMasker
```

**Comment by Remi-Gau** [comment]
Random thought: regarding loading of data to a SurfaceImage object.
Currently when instantiating, the API requires to pass both data and mesh as parameters.
I am wondering if we could allow to use default mesh if None is passed as mesh: for example defaulting to a freesurfer mesh it that's a case (which is what we do in all our examples so far).
It would simplify things for users who want run a surface based analysis when they have volumetric data.

**Comment by Remi-Gau** [comment]
I would have gone with pial as default.
And the SurfaceImage object comprises both left and right hemisphere, so we should be good on this.

I will put that on the list of points to discuss at our internal hackathon next week because maybe we want to first release SurfaceImage that does not do too much at first and improve it in subsequent releases.

**Comment by bthirion** [comment]
I was just meaning that you cannot simply drop a Numpy array and expect nilearn will make a nice plot out of it. You need at least to specify what hemisphere you're considering.

**Comment by Remi-Gau** [comment]
- see https://github.com/nilearn/nilearn/pull/4560

**Comment by Remi-Gau** [comment]
Will extract some of the TODOs from the top message into separate issues for better visibility on what tasks are left.
